pkg/compactor/block_upload_test.go

```diff
@@ -1259,12 +1237,123 @@ func TestMultitenantCompactor_HandleBlockUpload_Complete(t *testing.T) {
 			default:
 				assert.Equal(t, http.StatusOK, resp.StatusCode)
 				assert.Empty(t, string(body))
+				exists, err := bkt.Exists(context.Background(), path.Join(tc.blockID, block.MetaFilename))
+				require.NoError(t, err)
+				require.True(t, exists)
 			}
+		})
+	}
+}
 
-			bkt.AssertExpectations(t)
+func TestMultitenantCompactor_ValidateAndComplete(t *testing.T) {
+	const tenantID = "test"
+	const blockID = "01G3FZ0JWJYJC0ZM6Y9778P6KD"
+	injectedError := fmt.Errorf("injected error")
 
-			if tc.verifyUpload != nil {
-				tc.verifyUpload(t, &bkt, tc.expMeta)
+	uploadingMetaPath := path.Join(tenantID, blockID, uploadingMetaFilename)
+	validationPath := path.Join(tenantID, blockID, validationFilename)
+	metaPath := path.Join(tenantID, blockID, block.MetaFilename)
+
+	validationSucceeds := func(_ context.Context) error { return nil }
+
+	testCases := []struct {
+		name                        string
+		errorInjector               func(op bucket.Operation, name string) error
+		validation                  func(context.Context) error
+		expectNoValidationFile      bool
+		expectErrorInValidationFile bool
+		expectMeta                  bool
+	}{
+		{
+			name:                        "validation fails",
+			validation:                  func(_ context.Context) error { return injectedError },
+			expectErrorInValidationFile: true,
+		},
+		{
+			name:          "validation fails, uploading error fails",
```

Comment:
What are we checking for in this test? (No "expect" field is set)

Reply:
The answer here goes along with the Nit about default values. It's checking that there is a validation file present and that there isn't an error present in it. I'll make this clearer.

Reply:
Went through and made each test clear. I like it much more now, thanks.